crypto/x509/v3_purp.c

```diff
@@ -550,19 +587,37 @@ int x509v3_cache_extensions(X509 *x)
         x->ex_flags |= EXFLAG_INVALID;
 #endif
     for (i = 0; i < X509_get_ext_count(x); i++) {
-        ex = X509_get_ext(x, i);
-        if (OBJ_obj2nid(X509_EXTENSION_get_object(ex))
-            == NID_freshest_crl)
+        X509_EXTENSION *ex = X509_get_ext(x, i);
+        int nid = OBJ_obj2nid(X509_EXTENSION_get_object(ex));
+
+        if (nid == NID_freshest_crl)
             x->ex_flags |= EXFLAG_FRESHEST;
         if (!X509_EXTENSION_get_critical(ex))
             continue;
         if (!X509_supported_extension(ex)) {
             x->ex_flags |= EXFLAG_CRITICAL;
             break;
         }
+        switch (nid) {
+        case NID_basic_constraints:
+            x->ex_flags |= EXFLAG_BCONS_CRITICAL;
+            break;
+        case NID_authority_key_identifier:
+            x->ex_flags |= EXFLAG_AKID_CRITICAL;
+            break;
+        case NID_subject_key_identifier:
+            x->ex_flags |= EXFLAG_SKID_CRITICAL;
+            break;
+        case NID_subject_alt_name:
+            x->ex_flags |= EXFLAG_SAN_CRITICAL;
+            break;
+        default:
+            break;
+        }
     }
 
-    x509_init_sig_info(x);
+    /* Set x->siginf, ignoring errors due to unsupported algos */
+    (void)x509_init_sig_info(x);
```

Comment:
I'm don't understand why you make x509_init_sig_info() return an error, and then have the only caller ignore the error. I'm also unsure why we need x509_init_sig_info(), since it seems to just call x509_sig_info_init(), which has the same words in the name but in a different order. And x509_sig_info_init() also changed from returning void to an int, and also only has 1 caller.

Reply:
> I'm don't understand why you make x509_init_sig_info() return an error, and then have the only caller ignore the error.

Originally I wanted to make use of the error, but it turned out better not to do.
Yet for some other/future use of `x509_init_sig_info()` it may be useful to get an error returned.

>  I'm also unsure why we need x509_init_sig_info(), since it seems to just call x509_sig_info_init(), which has the same words in the name but in a different order. And x509_sig_info_init() also changed from returning void to an int, and also only has 1 caller.

I did not design this, but apparently the idea was that `x509_sig_info_init()` is more general such that it could be used also for handling other structures. For instance, in `crypto/x509/` there are other such cases where part of X509 and X509_CRL structures are treated uniformly.